lib/popcorn/pop_routes.nit

```diff
@@ -0,0 +1,216 @@
+# This file is part of NIT ( http://www.nitlanguage.org ).
+#
+# Copyright 2016 Alexandre Terrasa <alexandre@moz-code.org>
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Internal routes representation.
+module pop_routes
+
+import nitcorn
+intrude import nitcorn::vararg_routes
+
+# AppRoute provide services for path and uri manipulation and matching..
+#
+# Default strict like `/` or `/user` match strict the same URI string.
+# An exception is done for the trailing `/`, which is always omitted during the
+# parsing.
+#
+# ~~~
+# var route = new AppRoute("/")
+# assert route.match("")
+# assert route.match("/")
+# assert not route.match("/user")
+#
+# route = new AppRoute("/user")
+# assert not route.match("/")
+# assert route.match("/user")
+# assert route.match("/user/")
+# assert not route.match("/user/10")
+# assert not route.match("/foo")
+# ~~~
+class AppRoute
+
+	# Route relative path from server root.
+	var path: String
+
+	# Does self match the `req`?
+	fun match(uri: String): Bool do
+		uri = uri.simplify_path
+		var path = resolve_path(uri)
+		if uri.is_empty and path == "/" then return true
+		return uri == path
+	end
+
+	# Replace path parameters with concrete values from the `uri`.
+	#
+	# For strict routes, it returns the path unchanged:
+	# ~~~
+	# var route = new AppRoute("/")
+	# assert route.resolve_path("/user/10/profile") == "/"
+	#
+	# route = new AppRoute("/user")
+	# assert route.resolve_path("/user/10/profile") == "/user"
+	# ~~~
+	fun resolve_path(uri: String): String do return path.simplify_path
+
+	# Remove `resolved_path` prefix from `uri`.
+	#
+	# Mainly used to resolve and match mountable routes.
+	#
+	# ~~~
+	# var route = new AppRoute("/")
+	# assert route.uri_root("/user/10/profile") == "/user/10/profile"
+	#
+	# route = new AppRoute("/user")
+	# assert route.uri_root("/user/10/profile") == "/10/profile"
+	# ~~~
+	fun uri_root(uri: String): String do
+		var path = resolve_path(uri)
+		if path == "/" then return uri
+		return uri.substring(path.length, uri.length).simplify_path
+	end
+end
+
+# Parameterizable routes.
+#
+# Routes that can contains variables parts that will be resolved during the
+# matching process.
+#
+# Route parameters are marked with a colon `:`
+# ~~~
+# var route = new AppParamRoute("/:id")
+# assert not route.match("/")
+# assert route.match("/user")
+# assert route.match("/user/")
+# assert not route.match("/user/10")
```

Comment:
If you talk about `/:id`, yes `/username` will be matched because it's considered as an URI parameter.

For the route `/user`, no `/username` will not be matched. I added a test to show that in the strict routes.